### PR TITLE
Added Model for Malaria Detection #103

### PR DESCRIPTION
**Pull request by kritikaparmar-programmer**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/HakinCodes/Malaria-Detection/blob/master/CONTRIBUTING.md)?

Yes

### Description

Builded the model for Malaria Detection

### Checklist

- [x] I've read the contribution guidelines.
- [x] This PR mets all the requirements like Clean Code, proper README and Documentation

### Related Issues or Pull Requests

Fixes: #103 

### Screenshots (if any)

model

### CODE REVIEW

**Comment by HarshCasper** [comment]
Great work done @kritikaparmar-programmer 

Please add apt Jupyter Markdown over the Cells to briefly explain what your Code does. In the meantime, I would request other reviewers to review it.

**Comment by kritikaparmar-programmer** [comment]
> Great work done @kritikaparmar-programmer
> 
> Please add apt Jupyter Markdown over the Cells to briefly explain what your Code does. In the meantime, I would request other reviewers to review it.

Okk 👍

**Review by bhadreshpsavani — CHANGES_REQUESTED** [review]
Really impressive work @kritikaparmar-programmer , 

I have suggested a few changes,

1. We can have all of our imports in the first cell. 

2. We also remove imports like `classification_report` since we are not using it.

3. In the line,
```
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size = 0.2, random_state = 0)
```
We can make it like
```
X_train, X_valid, y_train, y_valid = train_test_split(data, labels, test_size = 0.2, random_state = 0)
```
I think, name it as validation set will be more appropriate since we already have holdout set as a test set

Note: You also need to make changes in the further steps according, since we are renaming the variable here.

**Comment by kritikaparmar-programmer** [comment]
Hey @HarshCasper  @bhadreshpsavani I've made the the changes Kindly review them...

**Review by bhadreshpsavani — CHANGES_REQUESTED** [review]
Hi @kritikaparmar-programmer 

There is a typo in Heading, `The Nueral Network gave Acuuracy of 94.99% ` It should be 
 `The Neural Network gave Accuracy of 94.99% `.

Can you please do this `..`  convey any special meaning or it was by mistake?
image

**Comment by kritikaparmar-programmer** [comment]
> Hi @kritikaparmar-programmer
> 
> There is a typo in Heading, `The Nueral Network gave Acuuracy of 94.99% ` It should be
> `The Neural Network gave Accuracy of 94.99% `.
> 
> Can you please do this `..` convey any special meaning or it was by mistake?
> image

Actually its my habit.. Sorry for that..😅

**Comment by kritikaparmar-programmer** [comment]
Hey @bhadreshpsavani Resolved the mistakes.

**Review by bhadreshpsavani — COMMENTED** [review]
Thanks for implementing changes really quickly @kritikaparmar-programmer
I am approving the pull request

**Review by bhadreshpsavani — CHANGES_REQUESTED** [review]
Thanks for implementing changes really quickly @kritikaparmar-programmer
I am approving the pull request

**Review by bhadreshpsavani — APPROVED** [review]
Thanks for implementing changes really quickly @kritikaparmar-programmer
I am approving the pull request

**Review by macabdul9 — CHANGES_REQUESTED** [review]
Great job @kritikaparmar-programmer.

I am requesting few changes that will make your notebook a bit elegant and fix a minor issue. 

1. Use Keras **data loaders** (to load the data and run **model.fit_generator** to train the model on generator object)  and apply **data augmentation** on training data as well. It will help to make our model robust and generalized (might be good for out of distribution samples and likely to enhance the performance) Eg: If you test your model on an inverted image from test data it might fail.

2. You have **2 neurons in the output layer** and you're using Sigmoid activation which maps values between [0, 1], outputs are not a probability distribution (sum of outputs will not be 1 which is what we need in order to compute CategoricalCrossentorpy/BinaryCrossentropy), so there might be cases where both outputs will be close to 0 or close to 1 or one close to 1 other close to 0, vice-versa, but in actual if one output is **p** other output should be **1-p** (and p should always be >= 0). Hence, actual accuracy may not be ~94% but need not to be discouraged, you've done a great job, here are the possible solutions (any one of them will fix the problem) you can apply to fix this problem.

> - Use **Softmax** activation in output layer rather than **Sigmoid** 
> - You can use single neuron in output layer with **Sigmoid activation** and apply MSE or BCE as loss function, **p < 0.5 = 0 else 1**
> - You can use tanh activation and MSE loss, if **p < 0 = class 0 else class 1**



Hope this helps.

**Comment by kritikaparmar-programmer** [comment]
> Great job @kritikaparmar-programmer.
> 
> I am requesting few changes that will make your notebook a bit elegant and fix a minor issue.
> 
> 1. Use Keras **data loaders** (to load the data and run **model.fit_generator** to train the model on generator object)  and apply **data augmentation** on training data as well. It will help to make our model robust and generalized (might be good for out of distribution samples and likely to enhance the performance) Eg: If you test your model on an inverted image from test data it might fail.
> 2. You have **2 neurons in the output layer** and you're using Sigmoid activation which maps values between [0, 1], outputs are not a probability distribution (sum of outputs will not be 1 which is what we need in order to compute CategoricalCrossentorpy/BinaryCrossentropy), so there might be cases where both outputs will be close to 0 or close to 1 or one close to 1 other close to 0, vice-versa, but in actual if one output is **p** other output should be **1-p** (and p should always be >= 0). Hence, actual accuracy may not be ~94% but need not to be discouraged, you've done a great job, here are the possible solutions (any one of them will fix the problem) you can apply to fix this problem.
> 
> > * Use **Softmax** activation in output layer rather than **Sigmoid**
> > * You can use single neuron in output layer with **Sigmoid activation** and apply MSE or BCE as loss function, **p < 0.5 = 0 else 1**
> > * You can use tanh activation and MSE loss, if **p < 0 = class 0 else class 1**
> 
> Hope this helps.

Hey, Can I do it like adding the data augmentation separately and not changing the previous one.. Like I can add the data augmentation part just below the simple model and later we can compare them also.. Tell me if we can do it in that way.

**Comment by macabdul9** [comment]
> > Great job @kritikaparmar-programmer.
> > I am requesting few changes that will make your notebook a bit elegant and fix a minor issue.
> > 
> > 1. Use Keras **data loaders** (to load the data and run **model.fit_generator** to train the model on generator object)  and apply **data augmentation** on training data as well. It will help to make our model robust and generalized (might be good for out of distribution samples and likely to enhance the performance) Eg: If you test your model on an inverted image from test data it might fail.
> > 2. You have **2 neurons in the output layer** and you're using Sigmoid activation which maps values between [0, 1], outputs are not a probability distribution (sum of outputs will not be 1 which is what we need in order to compute CategoricalCrossentorpy/BinaryCrossentropy), so there might be cases where both outputs will be close to 0 or close to 1 or one close to 1 other close to 0, vice-versa, but in actual if one output is **p** other output should be **1-p** (and p should always be >= 0). Hence, actual accuracy may not be ~94% but need not to be discouraged, you've done a great job, here are the possible solutions (any one of them will fix the problem) you can apply to fix this problem.
> > 
> > > * Use **Softmax** activation in output layer rather than **Sigmoid**
> > > * You can use single neuron in output layer with **Sigmoid activation** and apply MSE or BCE as loss function, **p < 0.5 = 0 else 1**
> > > * You can use tanh activation and MSE loss, if **p < 0 = class 0 else class 1**
> > 
> > 
> > Hope this helps.
> 
> Hey, Can I do it like adding the data augmentation separately and not changing the previous one.. Like I can add the data augmentation part just below the simple model and later we can compare them also.. Tell me if we can do it in that way.

Sure you can do, you can make these changes below the existing, but remember there is a little problem in the model that you have built [Issue 2 ] fix that as well. Also here are samples of using **[Example of using .flow_from_directory(directory)](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/ImageDataGenerator)** you can get rid of lots of code using this:
- Create **DataGenerator** (that's where you define augmentation, how you're going to read your images)
- Read the images using **flow_from_directory**, it's quite simple.

**Note: 2 Issue is important fixing that is mandatory**

**Comment by kritikaparmar-programmer** [comment]
Hey @macabdul9 I've done the changes . Kindly review them...

**Review by macabdul9 — APPROVED** [review]
